`TestAuth_RegisterUsingToken ` flakiness (data race)

## Failure

#### Link(s) to logs

- https://github.com/gravitational/teleport/actions/runs/5083826195/jobs/9135318256

#### Relevant snippet

```
==================
WARNING: DATA RACE
Write at 0x00c001721710 by goroutine 12263:
  github.com/gravitational/teleport/lib/auth.TestAuth_RegisterUsingToken()
      /__w/teleport/teleport/lib/auth/join_test.go:246 +0x2816
  testing.tRunner()
      /opt/go/src/testing/testing.go:1576 +0x216
  testing.(*T).Run.func1()
      /opt/go/src/testing/testing.go:1629 +0x47

Previous read at 0x00c001721710 by goroutine 12434:
  github.com/gravitational/teleport/lib/auth.TestAuth_RegisterUsingToken.func2.1()
      /__w/teleport/teleport/lib/auth/join_test.go:257 +0x56
  github.com/stretchr/testify/assert.Eventually.func1()
      /go/pkg/mod/github.com/stretchr/testify@v1.8.3/assert/assertions.go:1852 +0x39

Goroutine 12263 (running) created at:
  testing.(*T).Run()
      /opt/go/src/testing/testing.go:1629 +0x805
  testing.runTests.func1()
      /opt/go/src/testing/testing.go:2036 +0x8d
  testing.tRunner()
      /opt/go/src/testing/testing.go:1576 +0x216
  testing.runTests()
      /opt/go/src/testing/testing.go:2034 +0x87c
  testing.(*M).Run()
      /opt/go/src/testing/testing.go:1906 +0xb44
  github.com/gravitational/teleport/lib/auth.TestMain()
      /__w/teleport/teleport/lib/auth/auth_test.go:199 +0x3d
  main.main()
      _testmain.go:638 +0x33d

Goroutine 12434 (running) created at:
  github.com/stretchr/testify/assert.Eventually()
      /go/pkg/mod/github.com/stretchr/testify@v1.8.3/assert/assertions.go:1852 +0x3a5
  github.com/stretchr/testify/require.Eventually()
      /go/pkg/mod/github.com/stretchr/testify@v1.8.3/require/require.go:401 +0xb8
  github.com/gravitational/teleport/lib/auth.TestAuth_RegisterUsingToken.func2()
      /__w/teleport/teleport/lib/auth/join_test.go:256 +0x333
  testing.tRunner()
      /opt/go/src/testing/testing.go:1576 +0x216
  testing.(*T).Run.func1()
      /opt/go/src/testing/testing.go:1629 +0x47
==================
    testing.go:1446: race detected during execution of test
--- FAIL: TestAuth_RegisterUsingToken (0.37s)
===================================================
OUTPUT github.com/gravitational/teleport/lib/auth.TestAuth_RegisterUsingToken/reject_expired_dynamic_token
===================================================
=== RUN   TestAuth_RegisterUsingToken/reject_expired_dynamic_token
{"caller":"auth/join.go:106","component":"auth","level":"info","message":"Node \"node-name\" [localhost] is trying to join with role: Node.","timestamp":"2023-05-25T19:36:33Z"}
{"caller":"auth/join.go:52","component":"auth","level":"warning","message":"\"node-name\" can not join the cluster with role Node, token error: token expired or not found","timestamp":"2023-05-25T19:36:33Z"}
{"caller":"lite/periodic.go:49","component":"sqlite","level":"debug","message":"Failed to run remove expired keys: database is in readonly mode","timestamp":"2023-05-25T19:36:33Z"}
    join_test.go:256: 
        	Error Trace:	/__w/teleport/teleport/lib/auth/join_test.go:256
        	Error:      	Condition never satisfied
        	Test:       	TestAuth_RegisterUsingToken/reject_expired_dynamic_token
--- FAIL: TestAuth_RegisterUsingToken/reject_expired_dynamic_token (0.10s)
===================================================
```

--- a/lib/auth/join_test.go
+++ b/lib/auth/join_test.go
@@ -23,7 +23,6 @@ import (
 	"time"
 
 	"github.com/gravitational/trace"
-	"github.com/jonboulle/clockwork"
 	"github.com/stretchr/testify/require"
 	"golang.org/x/crypto/ssh"
 
@@ -43,7 +42,6 @@ func TestAuth_RegisterUsingToken(t *testing.T) {
 	ctx := context.Background()
 	p, err := newTestPack(ctx, t.TempDir())
 	require.NoError(t, err)
-	a := p.a
 
 	// create a static token
 	staticToken := types.ProvisionTokenV1{
@@ -57,12 +55,22 @@ func TestAuth_RegisterUsingToken(t *testing.T) {
 	err = p.a.SetStaticTokens(staticTokens)
 	require.NoError(t, err)
 
-	// create a dynamic token
+	// create a valid dynamic token
 	dynamicToken := generateTestToken(
 		ctx,
 		t,
-		types.SystemRoles{types.RoleNode}, time.Now().Add(time.Minute*30),
-		a,
+		types.SystemRoles{types.RoleNode},
+		p.a.GetClock().Now().Add(time.Minute*30),
+		p.a,
+	)
+
+	// create an expired dynamic token
+	expiredDynamicToken := generateTestToken(
+		ctx,
+		t,
+		types.SystemRoles{types.RoleNode},
+		p.a.GetClock().Now().Add(-time.Minute*30),
+		p.a,
 	)
 
 	sshPrivateKey, sshPublicKey, err := testauthority.New().GenerateKeyPair()
@@ -76,7 +84,6 @@ func TestAuth_RegisterUsingToken(t *testing.T) {
 		req              *types.RegisterUsingTokenRequest
 		certsAssertion   func(*proto.Certs)
 		errorAssertion   func(error) bool
-		clock            clockwork.Clock
 		waitTokenDeleted bool // Expired tokens are deleted in background, might need slight delay in relevant test
 	}{
 		{
@@ -216,7 +223,7 @@ func TestAuth_RegisterUsingToken(t *testing.T) {
 		{
 			desc: "reject expired dynamic token",
 			req: &types.RegisterUsingTokenRequest{
-				Token:        dynamicToken,
+				Token:        expiredDynamicToken,
 				HostID:       "localhost",
 				NodeName:     "node-name",
 				Role:         types.RoleNode,
@@ -224,37 +231,31 @@ func TestAuth_RegisterUsingToken(t *testing.T) {
 				PublicTLSKey: tlsPublicKey,
 			},
 			waitTokenDeleted: true,
-			clock:            clockwork.NewFakeClockAt(time.Now().Add(time.Hour + 1)),
 			errorAssertion:   trace.IsAccessDenied,
 		},
 		{
 			// relies on token being deleted during previous testcase
 			desc: "expired token should be gone",
 			req: &types.RegisterUsingTokenRequest{
-				Token:        dynamicToken,
+				Token:        expiredDynamicToken,
 				HostID:       "localhost",
 				NodeName:     "node-name",
 				Role:         types.RoleNode,
 				PublicSSHKey: sshPublicKey,
 				PublicTLSKey: tlsPublicKey,
 			},
-			clock:          clockwork.NewRealClock(),
 			errorAssertion: trace.IsAccessDenied,
 		},
 	}
 
 	for _, tc := range testcases {
 		t.Run(tc.desc, func(t *testing.T) {
-			if tc.clock == nil {
-				tc.clock = clockwork.NewRealClock()
-			}
-			a.SetClock(tc.clock)
-			certs, err := a.RegisterUsingToken(ctx, tc.req)
+			certs, err := p.a.RegisterUsingToken(ctx, tc.req)
 			if tc.errorAssertion != nil {
 				require.True(t, tc.errorAssertion(err))
 				if tc.waitTokenDeleted {
 					require.Eventually(t, func() bool {
-						_, err := a.ValidateToken(ctx, tc.req.Token)
+						_, err := p.a.ValidateToken(ctx, tc.req.Token)
 						return err != nil && strings.Contains(err.Error(), TokenExpiredOrNotFound)
 					}, time.Millisecond*100, time.Millisecond*10)
 				}

